src/mlpack/methods/decision_tree/best_binary_numeric_split_impl.hpp

```diff
@@ -185,14 +184,147 @@ double BestBinaryNumericSplit<FitnessFunction>::SplitIfBetter(
   return bestFoundGain;
 }
 
+template<typename FitnessFunction>
+template<bool UseWeights, typename VecType, typename WeightVecType>
+double BestBinaryNumericSplit<FitnessFunction>::SplitIfBetter(
+    const double bestGain,
+    const VecType& data,
+    const arma::Row<double>& labels,
+    const WeightVecType& weights,
+    const size_t minimumLeafSize,
+    const double minimumGainSplit,
+    double& splitInfo,
+    AuxiliarySplitInfo& /* aux */)
+{
+  // First sanity check: if we don't have enough points, we can't split.
+  if (data.n_elem < (minimumLeafSize * 2))
+    return DBL_MAX;
+  if (bestGain == 0.0)
+    return DBL_MAX; // It can't be outperformed.
+
+  // Next, sort the data.
+  arma::uvec sortedIndices = arma::sort_index(data);
+  arma::Row<double> sortedLabels(labels.n_elem);
+  arma::rowvec sortedWeights;
+  for (size_t i = 0; i < sortedLabels.n_elem; ++i)
+    sortedLabels[i] = labels[sortedIndices[i]];
+
+  // Sanity check: if the first element is the same as the last, we can't split
+  // in this dimension.
+  if (data[sortedIndices[0]] == data[sortedIndices[sortedIndices.n_elem - 1]])
+    return DBL_MAX;
+
+  // Only initialize if we are using weights.
+  if (UseWeights)
+  {
+    sortedWeights.set_size(sortedLabels.n_elem);
+    // The weights must keep the same order as the labels.
+    for (size_t i = 0; i < sortedLabels.n_elem; ++i)
+      sortedWeights[i] = weights[sortedIndices[i]];
+  }
+
+  double bestFoundGain = std::min(bestGain + minimumGainSplit, 0.0);
+  bool improved = false;
+  // Force a minimum leaf size of 1 (empty children don't make sense).
+  const size_t minimum = std::max(minimumLeafSize, (size_t) 1);
+
+  double totalWeight = 0.0;
+  double totalLeftWeight = 0.0;
+  double totalRightWeight = 0.0;
+
+  if (UseWeights)
+  {
+    totalWeight = arma::accu(sortedWeights);
+    bestFoundGain *= totalWeight;
+
+    for (size_t i = 0; i < minimum - 1; ++i)
+      totalLeftWeight += sortedWeights[i];
+
+    for (size_t i = minimum - 1; i < data.n_elem; ++i)
+      totalRightWeight += sortedWeights[i];
+  }
+  else
+  {
+    bestFoundGain *= data.n_elem;
+  }
+
+  // Loop through all possible split points, choosing the best one.
+  for (size_t index = minimum; index < data.n_elem - minimum + 1; ++index)
+  {
+    if (UseWeights)
+    {
+      totalLeftWeight += sortedWeights[index - 1];
+      totalRightWeight -= sortedWeights[index - 1];
+    }
+    // Make sure that the value has changed.
+    if (data[sortedIndices[index]] == data[sortedIndices[index - 1]])
+      continue;
+
+    /* TODO: The following function calculates the gain for each split each time from scratch
+             This can be greatly improved using advanced techniques like prefix sum and
+             prefix sum of squares etc. This will have drastic effects on runtime and is
+             definitely something we would want in future.
+     */
```

Comment:
We have to handle it before we move on with XGBoost otherwise XGBoost will run horribly slow. This PR is getting way too big. I think once this PR gets merged, then I will open another one to handle this. This will require benchmarking etc., so we will deal with it after this 👍

Reply:
That sounds like a great plan to me. :+1: